src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionCommitManager.cs

```diff
@@ -0,0 +1,170 @@
+{
+    internal sealed class CompletionCommitManager : IAsyncCompletionCommitManager
+    {
+        private static readonly IEnumerable<char> s_commitChars = ImmutableArray.Create(
+            ' ', '{', '}', '[', ']', '(', ')', '.', ',', ':',
+            ';', '+', '-', '*', '/', '%', '&', '|', '^', '!',
+            '~', '=', '<', '>', '?', '@', '#', '\'', '\"', '\\');
+
+        private static readonly EditorCompletion.CommitResult CommitResultUnhandled = new EditorCompletion.CommitResult(isHandled: false, EditorCompletion.CommitBehavior.None);
+
+        IEnumerable<char> IAsyncCompletionCommitManager.PotentialCommitCharacters => s_commitChars;
+
+        public bool ShouldCommitCompletion(IAsyncCompletionSession session, SnapshotPoint location, char typedChar, CancellationToken cancellationToken)
+            => s_commitChars.Contains(typedChar);
+
+        public EditorCompletion.CommitResult TryCommit(IAsyncCompletionSession session, ITextBuffer subjectBuffer, EditorCompletion.CompletionItem item, char typeChar, CancellationToken cancellationToken)
+        {
+            var document = subjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return CommitResultUnhandled;
+            }
+
+            var completionService = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(CompletionItemSource.RoslynItem, out var roslynItem))
+            {
+                // This isn't an item we provided (e.g. Razor). Let the editor handle it normally.
+                return CommitResultUnhandled;
+            }
+
+            // We can be called before for ShouldCommitCompletion. However, that call does not provide rules applied for the completion item.
+            // Now we check for the commit charcter in the context of Rules that could change the list of commit characters.
+            if (!IsCommitCharacter(typeChar, roslynItem.Rules.CommitCharacterRules))
+            {
+                return new EditorCompletion.CommitResult(isHandled: true, EditorCompletion.CommitBehavior.CancelCommit);
+            }
+
+            if (completionService.GetProvider(roslynItem) is IFeaturesCustomCommitCompletionProvider featuresCustomCommitCompletionProvider)
+            {
+                var commitBehavior = CustomCommit(document, featuresCustomCommitCompletionProvider, session.TextView, subjectBuffer, roslynItem, session.ApplicableToSpan, typeChar, cancellationToken);
+                return new EditorCompletion.CommitResult(isHandled: true, commitBehavior);
+            }
+
+            if (document.Project.Language == LanguageNames.VisualBasic && typeChar == '\n')
+            {
+                return new EditorCompletion.CommitResult(isHandled: false, EditorCompletion.CommitBehavior.RaiseFurtherReturnKeyAndTabKeyCommandHandlers);
+            }
```

Comment:
this code should not be here.  this is a core layer that needs to be language neutral.  For example, if TypeScript needs tha tbehavior, there is no way for them to get it.

Reply:
[https://github.com/dotnet/roslyn/issues/30276](https://github.com/dotnet/roslyn/issues/30276)

---
In reply to: [222074485](https://github.com/dotnet/roslyn/pull/29016#discussion_r222074485) [](ancestors = 222074485)